src/runtime/builtin_modules/builtins.cpp

```diff
@@ -145,30 +145,68 @@ extern "C" Box* any(Box* container) {
     return boxBool(false);
 }
 
-extern "C" Box* min(Box* arg0, BoxedTuple* args) {
+Box* min_max(Box* arg0, BoxedTuple* args, BoxedDict* kwargs, int opid) {
     assert(args->cls == tuple_cls);
+    assert(kwargs->cls == dict_cls);
 
-    Box* minElement;
+    bool pass_kwargs = (kwargs && kwargs->d.size());
+    Box* key_func;
+    Box* extremElement;
     Box* container;
+    Box* extremVal;
+
+    if (pass_kwargs) {
+        static BoxedString* key_str = static_cast<BoxedString*>(PyString_InternFromString("key"));
+        auto it = kwargs->d.find(key_str);
+        key_func = it->second;
+    }
+
     if (args->size() == 0) {
-        minElement = nullptr;
+        extremElement = nullptr;
+        extremVal = nullptr;
         container = arg0;
     } else {
-        minElement = arg0;
+        extremElement = arg0;
+        if (pass_kwargs) {
+            extremVal = runtimeCall(key_func, ArgPassSpec(1), extremElement, NULL, NULL, NULL, NULL);
+        } else {
+            extremVal = extremElement;
+        }
         container = args;
     }
 
+    Box* curVal = nullptr;
     for (Box* e : container->pyElements()) {
-        if (!minElement) {
-            minElement = e;
+        if (pass_kwargs) {
+            curVal = runtimeCall(key_func, ArgPassSpec(1), e, NULL, NULL, NULL, NULL);
+            if (!extremElement) {
+                extremVal = runtimeCall(key_func, ArgPassSpec(1), e, NULL, NULL, NULL, NULL);
+                extremElement = e;
+            }
         } else {
-            int r = PyObject_RichCompareBool(minElement, e, Py_GT);
-            if (r == -1)
-                throwCAPIException();
-            if (r)
-                minElement = e;
+            if (!extremElement) {
+                extremVal = e;
+                extremElement = e;
+            }
+            curVal = e;
+        }
+        int r = PyObject_RichCompareBool(extremVal, curVal, opid);
```

Comment:
If `extremElement`, we probably shouldn't call PyObject_RichCompareBool.  Maybe we could take the `!extremElement` check out the above if/else, and if it's true then `continue`?